README.md
# kemba
[![License: MIT](https://img.shields.io/badge/License-MIT-brightgreen.svg)](https://github.com/clok/kemba/blob/master/LICENSE)
[![Go Report Card](https://goreportcard.com/badge/clok/kemba)](https://goreportcard.com/report/clok/kemba)
[![Coverage Status](https://coveralls.io/repos/github/clok/kemba/badge.svg)](https://coveralls.io/github/clok/kemba)
[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/github.com/clok/kemba?tab=overview)
[![Mentioned in Awesome
Go](https://awesome.re/mentioned-badge-flat.svg)](https://github.com/avelino/awesome-go)

`debug` logging tool inspired by https://github.com/visionmedia/debug

#### Why is it named `kemba`?

`debug` is more generally considered to be [`runtime/debug`](https://golang.org/pkg/runtime/debug/) within Go. Since this takes heavy inspiration from my experiences using the `npm` module [`debug`](https://github.com/visionmedia/debug) I wanted to find a word that was somewhat connected to the inspiration. According to [Google translate](https://www.google.com/search?q=debug+in+icelandic) "debug" in English translated to Icelandic results in "kemba".

## Usage

The `kemba` logger reads the `DEBUG` and `KEMBA` environment variables to determine if a log line should be output. The logger outputs to `STDERR`.

When it is not set, the logger will immediately return, taking no action.

When the value is set (ex. `DEBUG=example:*,tool:details` and/or `KEMBA=plugin:fxn:start`), the logger will determine if it should be `enabled` when instantiated.

The value of these flags can be a simple regex alternative where a wildcard (`*`) are replaced with `.*` and all terms are prepended with `^` and appended with `$`. If a term does not include a wildcard, then an exact match it required.

Example of a wildcard in the middle of a tag string: `DEBUG=example:*:fxn` will match tags like `[example:tag1:fxn, example:tag2:fxn, example:anything:fxn, ...]`

To disabled colors, set the `NOCOLOR` environment variable to any value.

![image](https://user-images.githubusercontent.com/1429775/88557149-7973ff80-cfef-11ea-8ec2-ff332fd1b25f.png)

```go
package main

import (
    "time"

	"github.com/clok/kemba"
)

type myType struct {
	a, b int
}

// When the DEBUG or KEMBA environment variable is set to DEBUG=example:* the kemba logger will output to STDERR
func main () {
    k := kemba.New("example:tag")
	
    var x = []myType{
    	{1, 2},
    	{3, 4},
    }
    k.Printf("%#v", x)
    // Output to os.Stderr
    // example:tag []main.myType{main.myType{a:1, b:2}, main.myType{a:3, b:4}} +0s

    // Artificial delay to demonstrate the time tagging
    time.Sleep(250 * time.Millisecond)
    k.Printf("%# v", x)
    k.Println(x)

    // Artificial delay to demonstrate the time tagging
    time.Sleep(100 * time.Millisecond)
    k.Log(x)
    // All result in the same output to os.Stderr
    // example:tag []main.myType{ +XXms
    // example:tag     {a:1, b:2},
    // example:tag     {a:3, b:4},
    // example:tag }

    // Create a new logger with an extended tag
    k1 := k.Extend("1")
    k1.Println("a string", 12, true)
    // Output to os.Stderr
    // example:tag:1 a string +0s
    // example:tag:1 int(12)
    // example:tag:1 bool(true)
}
```

## Development

1. Fork the [clok/kemba](https://github.com/clok/kemba) repo
1. Use `go >= 1.16`
1. Branch & Code
1. Run linters :broom: `golangci-lint run`
    - The project uses [golangci-lint](https://golangci-lint.run/usage/install/#local-installation)
1. Commit with a Conventional Commit
1. Open a PR

CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v1.2.0"></a>
## [v1.2.0] - 2022-06-10
### Features
- support mid tag wildcard for enabling logger ([#31](https://github.com/clok/kemba/issues/31))


<a name="v1.1.2"></a>
## [v1.1.2] - 2022-06-10
### Bug Fixes
- **deps:** update all non-major dependencies ([#30](https://github.com/clok/kemba/issues/30))
- **deps:** update module github.com/stretchr/testify to v1.7.1 ([#27](https://github.com/clok/kemba/issues/27))
- **deps:** update module github.com/gookit/color to v1.4.2 ([#21](https://github.com/clok/kemba/issues/21))

### Chore
- update changelog for v1.1.2
- create FUNDING.yml
- update README
- **deps:** update golangci/golangci-lint-action action to v3 ([#26](https://github.com/clok/kemba/issues/26))
- **deps:** update actions/setup-go action to v3 ([#25](https://github.com/clok/kemba/issues/25))
- **deps:** update actions/checkout action to v3 ([#24](https://github.com/clok/kemba/issues/24))
- **deps:** update all non-major dependencies ([#22](https://github.com/clok/kemba/issues/22))
- **deps:** update goreleaser/goreleaser-action action to v3 ([#28](https://github.com/clok/kemba/issues/28))
- **deps:** update jandelgado/gcov2lcov-action action to v1.0.9 ([#29](https://github.com/clok/kemba/issues/29))


<a name="v1.1.1"></a>
## [v1.1.1] - 2021-04-05
### Chore
- update changelog for v1.1.1

### Ci
- add makefile and go releaser
- **renovate:** fix syntax error in renovate config


<a name="v1.1.0"></a>
## [v1.1.0] - 2021-03-15
### Chore
- **ci:** port to using golangci-lint github action
- **go.mod:** bump to go 1.16

### Features
- **release:** v1.1.0
- **release:** v1.1.0


<a name="v1.0.0"></a>
## [v1.0.0] - 2021-03-03
### Chore
- **deps:** update actions/checkout action to v2 ([#11](https://github.com/clok/kemba/issues/11))
- **deps:** update module gookit/color to v1.3.0 ([#7](https://github.com/clok/kemba/issues/7))
- **deps:** update actions/setup-go action to v2 ([#12](https://github.com/clok/kemba/issues/12))
- **deps:** update module stretchr/testify to v1.7.0 ([#16](https://github.com/clok/kemba/issues/16))
- **deps:** update module gookit/color to v1.3.6 ([#15](https://github.com/clok/kemba/issues/15))
- **deps:** update jandelgado/gcov2lcov-action action to v1.0.8 ([#14](https://github.com/clok/kemba/issues/14))
- **deps:** update module gookit/color to v1.3.2 ([#13](https://github.com/clok/kemba/issues/13))
- **deps:** update module gookit/color to v1.3.1 ([#8](https://github.com/clok/kemba/issues/8))
- **deps:** update coverallsapp/github-action action to v1.1.2 ([#9](https://github.com/clok/kemba/issues/9))
- **deps:** update jandelgado/gcov2lcov-action action to v1.0.7 ([#10](https://github.com/clok/kemba/issues/10))
- **github actions:** add go proxy warming
- **lint:** fix linting error on struct
- **renovate:** add gomodTidy option
- **renovate:** add extension for group:allNonMajor

### Features
- **release:** v1.0.0


<a name="v0.7.1"></a>
## [v0.7.1] - 2020-08-21
### Bug Fixes
- return pointer to color object

### Features
- **release:** v0.7.1


<a name="v0.7.0"></a>
## [v0.7.0] - 2020-08-21
### Features
- export PickColor helper method
- **release:** v0.7.0


<a name="v0.6.4"></a>
## [v0.6.4] - 2020-08-21
### Chore
- **dependencies:** remove duplicate sync of gookit/color
- **deps:** update module kr/pretty to v0.2.1 ([#6](https://github.com/clok/kemba/issues/6))
- **deps:** update module gookit/color to v1.2.7 ([#4](https://github.com/clok/kemba/issues/4))
- **deps:** add renovate.json ([#3](https://github.com/clok/kemba/issues/3))
- **renovate:** update config
- **renovate:** clean up dupe config
- **renovate:** add config file

### Features
- **release:** v0.6.4


<a name="v0.6.3"></a>
## [v0.6.3] - 2020-08-02
### Chore
- update readme with new badges, including awesome-go

### Features
- **release:** v0.6.3


<a name="v0.6.2"></a>
## [v0.6.2] - 2020-07-29
### Chore
- updated docs on exported struct

### Features
- **release:** v0.6.2


<a name="v0.6.1"></a>
## [v0.6.1] - 2020-07-27
### Chore
- do not export base kemba struct

### Features
- **release:** v0.6.1


<a name="v0.6.0"></a>
## [v0.6.0] - 2020-07-27
### Features
- **release:** v0.6.0
- **time delta:** added log event time deltas to first line of logged buffer events. Resolution in ms


<a name="v0.5.0"></a>
## [v0.5.0] - 2020-07-21
### Features
- **env:** support reading both the DEBUG and KEMBA env vars
- **release:** v0.5.0


<a name="v0.4.0"></a>
## [v0.4.0] - 2020-07-21
### Chore
- updated README

### Features
- **color:** deterministic color picking based on tag input for a new logger
- **release:** v0.4.0


<a name="v0.3.0"></a>
## [v0.3.0] - 2020-07-19
### Features
- **extend:** capability to extend existing logger with appended tag values

### Release
- **v0.3.0:** update CHANGELOG


<a name="v0.2.3"></a>
## [v0.2.3] - 2020-07-17
### Chore
- updated README usage notes

### Release
- **v0.2.3:** update CHANGELOG


<a name="v0.2.2"></a>
## [v0.2.2] - 2020-07-17
### Chore
- updated README example code.

### Release
- **v0.2.2:** update CHANGELOG


<a name="v0.2.1"></a>
## [v0.2.1] - 2020-07-17
### Chore
- updates to docs

### Release
- **v0.2.1:** update CHANGELOG


<a name="v0.2.0"></a>
## [v0.2.0] - 2020-07-17
### Bug Fixes
- adjust go.mod package name

### Chore
- updated docs and README
- added github actions & test coverage ([#2](https://github.com/clok/kemba/issues/2))
- update readme with example output
- added changelog

### Release
- **v0.2.0:** update CHANGELOG

### Tech Debt
- reduced color list to more agreeable list of select colors


<a name="v0.1.0"></a>
## v0.1.0 - 2020-07-16
### Chore
- update README
- update README, add contribution guide and initial go.mod

### Features
- **logger:** initial implementation of Printf, Println and Log ([#1](https://github.com/clok/kemba/issues/1))
- **tag:** enable regex tag processing for filtering of log calls

### Initial
- commit



kemba.go
package kemba

import (
	"bufio"
	"bytes"
	"fmt"
	"github.com/gookit/color"
	"github.com/kr/pretty"
	"hash/crc64"
	"log"
	"math/rand"
	"os"
	"regexp"
	"strings"
	"time"
)

// Kemba is a container struct for the Kemba logger.
// It used to manage the state of the logger.
// Currently all properties are not exported.
type Kemba struct {
	tag     string
	allowed string
	enabled bool
	logger  *log.Logger
	color   bool
	last    time.Time
}

var (
	table  = crc64.MakeTable(crc64.ISO)
	gs     = color.C256(uint8(240))
	colors = []int{
		20,
		21,
		26,
		27,
		32,
		33,
		38,
		39,
		40,
		41,
		42,
		43,
		44,
		45,
		56,
		57,
		62,
		63,
		68,
		69,
		74,
		75,
		76,
		77,
		78,
		79,
		80,
		81,
		92,
		93,
		98,
		99,
		112,
		113,
		128,
		129,
		134,
		135,
		148,
		149,
		160,
		161,
		162,
		163,
		164,
		165,
		166,
		167,
		168,
		169,
		170,
		171,
		172,
		173,
		178,
		179,
		184,
		185,
		196,
		197,
		198,
		199,
		200,
		201,
		202,
		203,
		204,
		205,
		206,
		207,
		208,
		209,
		214,
		215,
		220,
		221,
	}
)

// New Returns a Kemba logging instance. It will determine if the logger should
// bypass logging actions or be activated.
func New(tag string) *Kemba {
	allowed := getDebugFlagFromEnv()

	logger := Kemba{tag: tag, allowed: allowed}

	if logger.allowed != "" {
		logger.enabled = determineEnabled(tag, allowed)
		logger.color = os.Getenv("NOCOLOR") == ""
	} else {
		logger.enabled = false
		logger.color = false
	}

	var prefix string
	if logger.enabled {
		if logger.color {
			s := PickColor(tag)
			prefix = s.Sprintf("%s ", tag)
		} else {
			prefix = fmt.Sprintf("%s ", tag)
		}

		logger.logger = log.New(os.Stderr, prefix, log.Lmsgprefix)
		logger.last = time.Now()
	}

	return &logger
}

// Printf is a convenience wrapper that will apply pretty.Formatter to the passed in variables.
//
// Calling Printf(f, x, y) is equivalent to fmt.Printf(f, pretty.Formatter(x), pretty.Formatter(y)).
func (k *Kemba) Printf(format string, v ...interface{}) {
	if k.enabled {
		elapsed := k.determineElapsed()

		var buf bytes.Buffer
		_, _ = pretty.Fprintf(&buf, format, v...)

		showDelta := true
		k.printBuffer(buf, elapsed, &showDelta)
	}
}

// Println is a convenience wrapper that will apply pretty.Formatter to the passed in variables.
//
// Calling Println(x, y) is equivalent to fmt.Println(pretty.Formatter(x), pretty.Formatter(y)),
// but each operand is formatted with "%# v".
func (k *Kemba) Println(v ...interface{}) {
	if k.enabled {
		showDelta := true
		for _, x := range v {
			elapsed := k.determineElapsed()

			var buf bytes.Buffer
			_, _ = pretty.Fprintf(&buf, "%# v", x)

			k.printBuffer(buf, elapsed, &showDelta)
		}
	}
}

// Log is an alias to Println
func (k *Kemba) Log(v ...interface{}) {
	k.Println(v...)
}

// Extend returns a new Kemba logger instance that has appended the provided tag to the original logger.
//
// New logger instance will have original `tag` value delimited with a `:` and appended with the new extended `tag` input.
//
// Example:
//     k := New("test:original)
//     k.Log("test")
//     ke := k.Extend("plugin")
//     ke.Log("test extended")
//
// Output:
//     test:original test
//     test:original:plugin test extended
func (k *Kemba) Extend(tag string) *Kemba {
	exTag := fmt.Sprintf("%s:%s", k.tag, tag)
	return New(exTag)
}

// PickColor will return the same color based on input string.
//
// We want to pick the same color for a given tag to ensure consistent output behavior.
func PickColor(tag string) *color.Color256 {
	// Generate an 8 byte checksum to pass into Rand.seed
	seed := crc64.Checksum([]byte(tag), table)
	rand.Seed(int64(seed))
	v := rand.Intn(len(colors) - 1)
	s := color.C256(uint8(colors[v]))
	return &s
}

// printBuffer will append the elapsed time delta to the first line of the provided buffer
// if the showDelta parameter is true. Otherwise, this method prints the buffer lines to STDERR
func (k *Kemba) printBuffer(buf bytes.Buffer, elapsed time.Duration, showDelta *bool) {
	s := bufio.NewScanner(&buf)
	for s.Scan() {
		if *showDelta {
			var ft string
			if k.color {
				ft = gs.Sprintf("+%s", elapsed.Truncate(time.Millisecond))
			} else {
				ft = fmt.Sprintf("+%s", elapsed.Truncate(time.Millisecond))
			}
			k.logger.Printf("%s %s\n", s.Text(), ft)
			*showDelta = false
		} else {
			k.logger.Print(s.Text())
		}
	}
}

// getDebugFlagFromEnv considers both the value of DEBUG and KEMBA env values
// to determine the resulting logging flags to pass to the loggers.
func getDebugFlagFromEnv() string {
	dEnv := os.Getenv("DEBUG")
	kEnv := os.Getenv("KEMBA")

	var s []string
	if dEnv != "" {
		s = append(s, dEnv)
	}
	if kEnv != "" {
		s = append(s, kEnv)
	}

	if len(s) > 1 {
		return strings.Join(s, ",")
	} else if len(s) > 0 {
		return s[0]
	}

	return ""
}

// determineElapsed will determine the time delta from between the last log event for this
// Kemba logger and return the elapsed time.
func (k *Kemba) determineElapsed() time.Duration {
	now := time.Now()
	elapsed := now.Sub(k.last)
	k.last = now

	return elapsed
}

// determineEnabled will check the value of DEBUG and KEMBA environment variables to generate regex to test against the tag
//
// If no * in string, then assume exact match
// Else
// It will split by , and perform
// It will, replace * with .*
func determineEnabled(tag string, allowed string) bool {
	var a bool
	for _, l := range strings.Split(allowed, ",") {
		if strings.Contains(l, "*") {
			reg := strings.ReplaceAll(l, "*", ".*")
			if !strings.HasPrefix(reg, "^") && !strings.HasPrefix(reg, "*") {
				reg = fmt.Sprintf("^%s", reg)
			}

			if !strings.HasSuffix(reg, "$") {
				reg = fmt.Sprintf("%s$", reg)
			}

			if !a {
				a, _ = regexp.Match(reg, []byte(tag))
			}
		} else if !a {
			a = l == tag
		}
	}
	return a
}
